src/Microsoft.OData.Core/JsonLight/ODataJsonLightServiceDocumentDeserializer.cs

```diff
@@ -312,48 +304,333 @@ private ODataServiceDocumentElement ReadServiceDocumentElement(PropertyAndAnnota
+                                }
+
+                                break;
+                        }
+                    }).ConfigureAwait(false);
             }
 
+            // URL and Name are mandatory
+            ValidateServiceDocumentElementHasRequiredProperty(JsonLightConstants.ODataServiceDocumentElementName, name);
+            ValidateServiceDocumentElementHasRequiredProperty(JsonLightConstants.ODataServiceDocumentElementUrlName, url);
+
+            ODataServiceDocumentElement serviceDocumentElement = CreateServiceDocumentElement(kind);
+
             if (serviceDocumentElement != null)
             {
                 serviceDocumentElement.Url = this.ProcessUriFromPayload(url[0]);
                 serviceDocumentElement.Name = name[0];
                 serviceDocumentElement.Title = title[0];
             }
 
-            this.JsonReader.ReadEndObject();
+            await this.JsonReader.ReadEndObjectAsync()
+                .ConfigureAwait(false);
+
+            return serviceDocumentElement;
+        }
+
+        /// <summary>
+        /// Creates a <see cref="ODataServiceDocumentElement"/>.
+        /// </summary>
+        /// <param name="kind">Property values for the "kind" property.</param>
+        /// <returns>A <see cref="ODataServiceDocumentElement"/> instance.</returns>
+        private static ODataServiceDocumentElement CreateServiceDocumentElement(string[] kind)
+        {
+            // If not specified its an entity set.
+            if (kind[0] == null)
+            {
+                return new ODataEntitySetInfo();
+            }
+
+            ODataServiceDocumentElement serviceDocumentElement = null;
+
+            if (kind[0].Equals(JsonLightConstants.ServiceDocumentEntitySetKindName, StringComparison.Ordinal))
+            {
+                serviceDocumentElement = new ODataEntitySetInfo();
+            }
+            else if (kind[0].Equals(JsonLightConstants.ServiceDocumentFunctionImportKindName, StringComparison.Ordinal))
+            {
+                serviceDocumentElement = new ODataFunctionImportInfo();
+            }
+            else if (kind[0].Equals(JsonLightConstants.ServiceDocumentSingletonKindName, StringComparison.Ordinal))
+            {
+                serviceDocumentElement = new ODataSingletonInfo();
+            }
 
             return serviceDocumentElement;
         }
+
+        /// <summary>
+        /// Validates that the specified property in the service document element is not repeated.
+        /// </summary>
+        /// <param name="property">The property.</param>
+        /// <param name="propertyValues">Property values for the specified property.</param>
+        private static void ValidateServiceDocumentElementHasNoRepeatedProperty(string property, string[] propertyValues)
+        {
+            Debug.Assert(propertyValues != null && propertyValues.Length > 0, $"{nameof(propertyValues)} != null && {nameof(propertyValues)}.Length > 0");

```

Comment:
Refer to the same from `ValidateServiceDocumentElementHasRequiredProperty`

Reply:
See my comment here https://github.com/OData/odata.net/pull/2322#discussion_r812578515